xwiki-commons-core/xwiki-commons-store/xwiki-commons-store-blob/xwiki-commons-store-blob-api/src/main/java/org/xwiki/store/blob/BlobStore.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.store.blob;
+
+import java.util.stream.Stream;
+
+import org.xwiki.stability.Unstable;
+
+/**
+ * A storage that allows storing blob data.
+ *
+ * @version $Id$
+ * @since 17.10.0RC1
+ */
+@Unstable
+public interface BlobStore
+{
+    /**
+     * @return the unique name of this blob store
+     */
+    String getName();
+
+    /**
+     * @return the hint of this blob store (e.g., "filesystem", "s3")
+     */
+    String getHint();
+
+    /**
+     * @return the properties used to create this store
+     */
+    BlobStoreProperties getProperties();
+
+    /**
+     * Get the blob with the given identifier.
+     *
+     * @param path the path of the blob to retrieve
+     * @return the blob with the given path
+     * @throws BlobStoreException if there cannot be a blob with the given path (e.g., because the path is too long)
+     */
+    Blob getBlob(BlobPath path) throws BlobStoreException;
+
+    /**
+     * List all descendant blobs under the given path prefix. The returned stream includes blobs at every depth below
+     * the provided path. The caller must close the returned stream after use.
+     *
+     * @param path the path prefix to search under
+     * @return a stream of the blobs under the given path
+     * @throws BlobStoreException if the listing operation fails
+     */
+    Stream<Blob> listDescendants(BlobPath path) throws BlobStoreException;
```

Comment:
I would have also expected an API to get only direct descendants, sounds like a common use case.